datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -32,13 +33,19 @@ use arrow::{
     record_batch::RecordBatch,
 };
 
-use crate::PhysicalExpr;
+use crate::{expressions, PhysicalExpr};
 use arrow::array::*;
 use arrow::buffer::{Buffer, MutableBuffer};
 use datafusion_common::ScalarValue;
 use datafusion_common::{DataFusionError, Result};
 use datafusion_expr::ColumnarValue;
 
+/// Size at which to use a Set rather than Vec for `IN` / `NOT IN`
+/// Value chosen to be consistent with Spark
+/// https://github.com/apache/spark/blob/4e95738fdfc334c25f44689ff8c2db5aa7c726f2/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L259-L265
+/// TODO: add switch codeGen in In_List
+static OPTIMIZER_INSET_THRESHOLD: usize = 10;
```

Comment:
I don't think we should follow Spark, but should use some benchmarking to find a good heuristic for this threshold (where it's faster to use a hash set).

The optimal value might be quite a bit higher at this moment.

Reply:
Agree!  i will post a bench result. By the way, without subquery implement, Is there some way easy to do?

Reply:
Cool analysis!
I am wondering if other data types make it a bit different, such as strings / utf8 arrays? I expect the conversion to be a bit slower there, because of the extra conversion/allocations needed.

Reply:
I have try to use cast 'u32.to_string' get same result. In my opinion, use string may cause more cost in `hash`.

Reply:
I meant to benchmark it on utf8 columns. The conversion from array -> `ScalarValue` for each string is extra expensive, as it copies / allocates new `String`s.

Reply:
<img width="653" alt="image" src="https://user-images.githubusercontent.com/37145547/162232125-28122bd8-64c7-410f-9996-2df4bf27f942.png">
Cool, you are right! @Dandandan Thanks!  utf8 columns threshold near 50.
Maybe i will Set it by type👍

Reply:
Amazing, thanks for the comparison. I would also be OK with changing it to some value like 30. Seems for both numeric as for utf8 data a good enough.
At some point we could further optimize the implementation (https://github.com/apache/arrow-datafusion/issues/2165) after this we can adjust to a (lower) value.

Reply:
Done

Reply:
> where array -> ScalarValue copies / allocates happened in code

FYI, (String) allocations happen here, when converting to a` ScalarValue`.

Reply:
❤️ @Dandandan  Thanks a lot for your info !